rfq/order.go

```diff
@@ -0,0 +1,320 @@
+			"(htlc=%v, channel_remit=%v)", err, htlc, channelRemit)
+
+		return &lndclient.InterceptedHtlcResponse{
+			Action: lndclient.InterceptorActionFail,
+		}, nil
+	}
+
+	log.Debug("HTLC complies with channel remit. Broadcasting accept " +
+		"event.")
+	acceptHtlcEvent := NewAcceptHtlcEvent(htlc, *channelRemit)
+	h.cfg.AcceptHtlcEvents <- acceptHtlcEvent
+
+	return &lndclient.InterceptedHtlcResponse{
+		Action: lndclient.InterceptorActionResume,
+	}, nil
+}
+
+// setupHtlcIntercept sets up HTLC interception.
+func (h *OrderHandler) setupHtlcIntercept(ctx context.Context) error {
+	// Intercept incoming HTLCs. This call passes the handleIncomingHtlc
+	// function to the interceptor. The interceptor will call this function
+	// in a separate goroutine.
+	err := h.cfg.HtlcInterceptor.InterceptHtlcs(ctx, h.handleIncomingHtlc)
+	if err != nil {
+		if fn.IsCanceled(err) {
+			return nil
+		}
+
+		return fmt.Errorf("unable to setup incoming HTLC "+
+			"interceptor: %w", err)
+	}
+
+	return nil
+}
+
+// mainEventLoop executes the main event handling loop.
+func (h *OrderHandler) mainEventLoop() {
+	log.Debug("Starting main event loop for order handler")
+
+	cleanupTicker := time.NewTicker(h.cfg.CleanupInterval)
+	defer cleanupTicker.Stop()
+
+	for {
+		select {
+		// Periodically clean up expired channel remits from our local
+		// cache.
+		case <-cleanupTicker.C:
+			log.Debug("Cleaning up any stale channel remits from " +
+				"the order handler")
+			h.cleanupStaleChannelRemits()
+
+		case <-h.Quit:
+			log.Debug("Received quit signal. Stopping negotiator " +
+				"event loop")
+			return
+		}
+	}
+}
+
+// Start starts the service.
+func (h *OrderHandler) Start() error {
+	var startErr error
+	h.startOnce.Do(func() {
+		log.Info("Starting subsystem: order handler")
+
+		// Start the main event loop in a separate goroutine.
+		h.Wg.Add(1)
+		go func() {
+			defer h.Wg.Done()
+
+			ctx, cancel := h.WithCtxQuitNoTimeout()
+			defer cancel()
+
+			startErr = h.setupHtlcIntercept(ctx)
+			if startErr != nil {
+				log.Errorf("Error setting up HTLC "+
+					"interception: %v", startErr)
+				return
+			}
+
+			h.mainEventLoop()
+		}()
+	})
+
+	return startErr
+}
+
+// RegisterChannelRemit registers a channel management remit. If a remit exists
+// for the channel SCID, it is overwritten.
+func (h *OrderHandler) RegisterChannelRemit(quoteAccept rfqmsg.Accept) error {
+	log.Debugf("Registering channel remit for SCID: %d",
+		quoteAccept.ShortChannelId())
+
+	channelRemit := NewChannelRemit(quoteAccept)
+
+	h.channelRemitsMtx.Lock()
+	defer h.channelRemitsMtx.Unlock()
+
+	h.channelRemits[channelRemit.Scid] = channelRemit
+
+	return nil
+}
+
+// FetchChannelRemit fetches a channel remit given a serialised SCID. If a
+// channel remit is not found, false is returned. Expired channel remits are
+// not returned and are removed from the cache.
+func (h *OrderHandler) FetchChannelRemit(scid SerialisedScid) (*ChannelRemit,
```

Comment:
nit: here (and some other places) we don't distinguish between expired vs. non-existent - I'm not sure why we  want to erase that info? At a levels at least vs. just external RPCs.

Reply:
In the `OrderHandler` my intention was to only have active channel remits in state. But I do think its important to be able to respond with a "quote expired" message. I'll think a bit about how to implement that, it might be something we want to do once we have better persistence in place.

We need to think about handling many short lived quotes and therefore many expired quotes. For example, how long do we keep them around? I'd like to address in a future PR if that makes sense to you too.